File: Python/binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @return {integer[]}
    def preorderTraversal(self, root):
        stk = []
        ans = []
        if root: stk.append(root)
        while stk:
            root = stk.pop()
            ans.append(root.val)
            if root.right: stk.append(root.right)
            if root.left: stk.append(root.left)
        return ans

File: C++/binary-tree-preorder-traversal.cpp
/**
 *  * Definition for a binary tree node.
 *   * struct TreeNode {
 *    *     int val;
 *     *     TreeNode *left;
 *      *     TreeNode *right;
 *       *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 *        * };
 *         */
class Solution {
public:
    std::vector<int> preorderTraversal(TreeNode* root) {
        std::stack<TreeNode*> stk;
        std::vector<int> ans;
        if(root) stk.push(root);
        while(!stk.empty()) {
            root = stk.top();
            ans.push_back(root->val);
            stk.pop();
            if(root->right) stk.push(root->right);
            if(root->left) stk.push(root->left);
        }
        return ans;
    }
};

